Turn off BOOST_IOSTREAMS_HAS_DINKUMWARE_FPOS for MSVC++

BillyONeal opened the pull request:
Boost is attempting to use a function not in the standard, seekpos(),
guarded by this setting. Looks like Boost can use the standards
conforming interface for MSVC++ instead. It should work going
back effectively "forever" as far as MSVC++ is concerned, but this
guard applies the change only to 2017. It's possible there is a better
control in boost.config I missed.

eldiener commented:
The question is how we should be targeting vc++ 14.1 ? I think we should use BOOST_MSVC from config, which holds the value of vc++'s  _MSC_VER. Using a recent macro to identify a version of the vc++'s  standard library seems very chancy to me.

BillyONeal commented:
`_MSC_VER` should be OK if you'd prefer it; we added `_MSVC_STL_VERSION` specifically at the request of Boost for separate version numbers between the compiler and library (as sometimes new compilers get used with old libraries).

eldiener commented:
I would not use _MSC_VER directly in your PR but specifically BOOST_MSVC. This is because some other compilers, which are not vc++, also define _MSC_VER whereas BOOST_MSVC means that lvc++ specifically is being used. Also are you really saying that newer versions of vc++ could be used with old versions of your C++ standard library ? I did not think that was possible in Microsoft distributions.

When you say that 'Boost can use the standards conforming interface for MSVC++ instead. It should work going back effectively "forever" as far as MSVC++ is concerned' do you mean that all older versions of VC++, with their distributed versions of the C++ standard library, have a standards conforming interface as far as this problem is concerned ? That would be strange as it would put in question the original reason for the vc++ specific code in this particular situation.

BillyONeal commented:
> Also are you really saying that newer versions of vc++ could be used with old versions of your C++ standard library ?

Yes, its a thing people do sometimes. Old libs with new compiler should work, old compiler with new libs is unsupported.

> do you mean that all older versions of VC++, with their distributed versions of the C++ standard library, have a standards conforming interface as far as this problem is concerned ?

Modulo bugs, yes. I only tested with 2017 though which is why I so guarded it that way.

>That would be strange as it would put in question the original reason for the vc++ specific code in this particular situation.

The Dinkumware code path is used for other Dinkumware licensees; they have this behavior of "call fsetpos followed by fseek" because the C standard says it's UB to call fsetpos on anything not returned by a call to fgetpos. (See C11, 7.21.9.3 "The fsetpos function"/2) But Dinkumware's caution there happens to be unnecessary for at least the current CRT implementation.

There are some bugs that happened with VS 2017 15.7 and earlier where directly comparing fpos instances with non-`streamoff` integer types created ambiguities (compiler didn't know whether to convert fpos into streamoff and widen other integer, or convert other integer into fpos and compare those) which I fixed in 15.8 by adding comparison operators between fpos and all other integral types. (Note that 15.8 isn't even in preview yet; I just fixed this in our sources last week and found out about Boost depending on this nonstandard seekoff() member I removed because we build Boost with the prerelease compiler every night)

BillyONeal commented:
> directly comparing fpos instances with non-streamoff integer types

Also note that the non-Dinkumware path in Boost already deals with this problem by directly casting the fpos to streamoff first.

eldiener commented:
The original code path is for Dinkumware rather than for vc++ specifically. If you are saying that vc++ always worked correctly with Dinkumware and did not need this one-off code path we could remove it for any version  of vc++ by just checking if BOOST_MSVC is defined. But to be more conservative than that, in order not to destroy the use of iostreams for past versions of vc++ other than the present one, I am fine with approving your PR as long as _MSVC_STL_VERSION is never used by any other compiler, which I assume and hope will be the case, without checking for BOOST_MSVC. To be really conservative though it might be best to check for both _MSVC_STL_VERSION and BOOST_MSVC as a reason to not use the code path. Can you change it appropriately to do that or do you want me to merge the PRT and do it myself ? BOOST_MSVC is defined whenever vc++ is the compiler and not defined otherwise.

BillyONeal commented:
I was just informed last night that there's a Boost.IOStreams test failing right now; let me investigate and I'll keep you posted. Thanks!

BillyONeal commented:
Turns out the test failure is because I broke basic_stringbuf::seekoff last week :/, so not related to this change. I'll make the change you requested to test BOOST_MSVC first as soon as I hear I haven't broken boost.iostream any more. :)

BillyONeal commented:
OK, confirmed that it's just because I made seekoff corrupt basic_stringbuf which I just fixed. (Yay, thank you boost iostreams tests for catching my mistake!) Updated PR with the additional test you requested.

MarcelRaad commented:
Could this be merged to master please? This is needed for VS 2017 Update 8 as of Preview 3.0.

garethsb-sony commented:
As the original author of the workarounds in boost/iostreams/positioning.hpp, I can confirm they absolutely were necessary back when I produced them in 2005, but that's a long time ago now, even for Boost. ;-)

BillyONeal commented:
I apologize on behalf of my predecessors? :P

BillyONeal commented:
Wait wait wait wait we cast it to long? WHY?!???! :(

alanbirtles commented:
Before VS2010 lots of ugly hacks were required for seeking past 2GB in std::fstream for similar reasons.

mnordie commented:
Any idea when this will be released?
Doesn't seem to be in 1.68.

jeking3 commented:
This should appear in 1.69.0 - I think it was supposed to be in 1.68.0 but wasn't merged in time.  It is in the pipeline now.

mwpowellhtx commented:
I'm receiving the same error in [iostreams/positioning.hpp](/boostorg/iostreams/blob/a3e973f3d49ce4f5469f83f759e528073b05726a/include/boost/iostreams/positioning.hpp#L96). Make sure y'all check and double check your regression before making that commit and/or merging, please.

mwpowellhtx commented:
@BillyONeal It's a good question, it is, but how do I know whether I have *"Dinkumware fpos"*?

Meanwhile, I am defining ``_SILENCE_FPOS_SEEKPOS_DEPRECATION_WARNING``, but letting the deprecated (returning zero) ``seekpos()`` linger just seems like asking for regression trouble.

BillyONeal commented:
@mwpowellhtx It would be more appropriate to call it "workaround broken Dinkumware behavior where they put file sizes into an int" :)

mwpowellhtx commented:
@BillyONeal Copy that; I understand. :+1: 

I installed the patch, BTW, and it appears to be working. I still have to "acknowledge" the issue, so to speak, but at least I can proceed forward.

ibrahimLearning commented:
@BillyONeal 
would tell me please how can I turn off Boost_IOstream_has_dinkkumware_fpos? I already install boost with pcl!
